lib/puppet_x/puppetlabs/strings/yard/handlers/puppet_4x_function_handler.rb

```diff
@@ -33,22 +52,73 @@ class Puppet4xFunctionHandler < YARD::Handlers::Ruby::Base
     # Then, traverse the children of the parameters, storing each identifier
     # in the list of method arguments.
     obj = MethodObject.new(function_namespace, name) do |o|
+    end
+
+    # overload_signatures is a list of lists of tuples or empty:
+    # overload_signatures = [
+    #   [ # First function dispatch arguments
+    #     # argument name, argument type
+    #     ['arg0', 'Variant[String,Array[String]]'
+    #     ['arg1', 'Optional[Type]']
+    #   ],
+    #   [ # Second function dispatch arguments
+    #     ['arg0', 'Variant[String,Array[String]]'
+    #     ['arg1', 'Optional[Type]']
+    #     ['arg2', 'Any']
+    #   ]
+    # ]
+    overload_signatures = []
+    statement.traverse do |node|
+      # Find all of the dispatch methods
+      if node.type == :ident and node.source == 'dispatch'
+        command = node.parent
+        do_block = command.jump :do_block
+        # If the command doesn't have a do_block we can't extract type info
+        if do_block == command
+          next
+        end
+        # Iterate through each of the children of the do block and build
+        # tuples of parameter names and parameter type signatures
+        do_block.children.first.children.each do |child|
+          overload_signatures <<= [extract_type_from_command(child)]
+        end
+      end
+    end
+
+    # If the overload_signatures list is empty because we couldn't find any
+    # dispatch blocks, then there must be one function named the same as the
+    # name of the function being created.
+    if overload_signatures.length == 0
       statement.traverse do |node|
-        if node.type == :params
-          node.traverse do |params|
-            if params.type == :ident
-              # FIXME: Replace nil with parameter types
-              method_arguments << [params[0], nil]
+        # Find the function definition with the same name as the puppet
+        # function being created.
+        if (node.type == :def and node.children.first.type == :ident and
+            node.children.first.source == obj.name.to_s)
+          signature = []
+          # Find its parameters. If they don't exist, fine
+          params = node.jump :params
+          break if params == node
+          params.traverse do |param|
+            if param.type == :ident
+              # The parameters of Puppet functions with no defined dispatch are
+              # as though they are Any type.
+              signature << [param[0], 'Any']
```

Comment:
A ruby parameter may be a splat (e.g. def foo(a, b, *c) which we take as variable number of arguments